### PR TITLE
[16.0][IMP-FIX] sale_order_type: Remove onchange_type_id method

### PR DESCRIPTION
**Pull request by pilarvargas-tecnativa**
Remove onchange_type_id method by passing fields to compute writable. 

cc @Tecnativa TT44380

@pedrobaeza @chienandalu  please review :)

### CODE REVIEW

File: sale_order_type/models/sale.py [hunk 1]
```diff
@@ -67,41 +75,70 @@ def _compute_sale_type_id(self):
...
             record.type_id = sale_type
 
-    @api.onchange("type_id")
-    def onchange_type_id(self):
-        # TODO: To be changed to computed stored readonly=False if possible in v14?
-        vals = {}
+    @api.depends("type_id")
+    def _compute_warehouse_id(self):
+        res = super()._compute_warehouse_id()
         for order in self:
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py]: Do this on all the computed. We don't want that this empties some fields if you empty the sales order type.

```suggestion
        for order in self.filtered("type_id"):
```

File: sale_order_type/models/sale.py [hunk 2]
```diff
@@ -25,6 +26,13 @@ class SaleOrder(models.Model):
         copy=True,
         check_company=True,
     )
+    picking_policy = fields.Selection(
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py:30]:
```suggestion
    # Fields converted to computed writable
    picking_policy = fields.Selection(
```

File: sale_order_type/models/sale.py [hunk 3]
```diff
@@ -67,41 +75,70 @@ def _compute_sale_type_id(self):
...
-            vals = {}
             if order_type.warehouse_id:
-                vals.update({"warehouse_id": order_type.warehouse_id})
+                order.warehouse_id = order_type.warehouse_id
+        return res
+
+    @api.depends("type_id")
+    def _compute_picking_policy(self):
+        if hasattr(super(), "_compute_picking_policy"):
+            return super()._compute_picking_policy()
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py]:
```suggestion
            res = super()._compute_picking_policy()
```

and `return res` at the end.

File: sale_order_type/models/sale.py [hunk 4]
```diff
@@ -151,3 +188,18 @@ def _prepare_invoice(self):
...
             res["sale_type_id"] = self.type_id.id
         return res
+
+
+class SaleOrderLine(models.Model):
+    _inherit = "sale.order.line"
+
+    route_id = fields.Many2one(compute="_compute_route_id", store=True, readonly=False)
+
+    @api.depends("order_id.type_id")
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py:205]: I think this field is not correctly resolved in the one2many. Maybe you need an extra `order_type_id` helper computed field.

File: sale_order_type/models/sale.py [hunk 5]
```diff
@@ -67,41 +75,73 @@ def _compute_sale_type_id(self):
...
+    @api.depends("type_id")
+    def _compute_picking_policy(self):
+        if hasattr(super(), "_compute_picking_policy"):
+            super()._compute_picking_policy()
+        for order in self.filtered("type_id"):
+            order_type = order.type_id
             if order_type.picking_policy:
-                vals.update({"picking_policy": order_type.picking_policy})
+                order.picking_policy = order_type.picking_policy
+        return
```
Comment (pedrobaeza) [inline sale_order_type/models/sale.py]: Don't put return if you don't return anything.

File: sale_order_type/tests/test_sale_order_type.py [hunk 6]
```diff
@@ -195,8 +194,7 @@ def test_invoice_without_partner(self):
 
     def test_sale_order_flow_route(self):
         order = self.create_sale_order()
-        order.type_id = self.sale_type_route.id
-        order.onchange_type_id()
+        order.type_id = self.sale_type_route
```
Comment (pedrobaeza) [inline sale_order_type/tests/test_sale_order_type.py]: Is this change needed?

**Review by pedrobaeza — APPROVED** [review]
Thank you very much!

**Review by chienandalu — APPROVED** [review]
Thanks! :+1:

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
Cancelled, as there's no confirmation of this working on e-commerce. @pilarvargas-tecnativa would you?

**Comment by pilarvargas-tecnativa** [comment]
> Cancelled, as there's no confirmation of this working on e-commerce. @pilarvargas-tecnativa would you?

I confirm it doesn't work in e-commerce. I'm debugging the whole flow to see at what point the new e-commerce flow appears but I'm afraid nothing more can be done in this PR to make it work in e-commerce, I'll confirm and modify the commit as soon as I complete the debug.

**Comment by pilarvargas-tecnativa** [comment]
@pedrobaeza the changes in the commit message have been made.

**Review by chienandalu — APPROVED** [review]

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge patch
